c_src/membrane_ffmpeg_video_filter_plugin/text_overlay.c

```diff
@@ -0,0 +1,208 @@
+  }
+
+  UnifexPayload payload_frame;
+  size_t payload_size = av_image_get_buffer_size(
+      filtered_frame->format, filtered_frame->width, filtered_frame->height, 1);
+  unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, payload_size,
+                       &payload_frame);
+
+  if (av_image_copy_to_buffer(payload_frame.data, payload_size,
+                              (const uint8_t *const *)filtered_frame->data,
+                              filtered_frame->linesize, filtered_frame->format,
+                              filtered_frame->width, filtered_frame->height,
+                              1) < 0) {
+    res = filter_result_error(env, "copy_to_payload");
+    goto exit_filter;
+  }
+  res = filter_result_ok(env, &payload_frame);
+exit_filter:
+  av_frame_free(&frame);
+  av_frame_free(&filtered_frame);
+  return res;
+}
+
+void handle_destroy_state(UnifexEnv *env, State *state) {
+  UNIFEX_UNUSED(env);
+  if (state->filter_graph != NULL) {
+    avfilter_graph_free(&state->filter_graph);
+  }
+}
+
+void create_filter_description(char *filter_descr, int len, char *text,
+                               int fontsize, int box, char *boxcolor,
+                               int border, char *fontcolor, char *fontfile,
+                               char *x, char *y) {
+
+  filter_descr += snprintf(filter_descr, len, "drawtext=text=%s", text);
+  if (fontsize != -1) {
+    filter_descr += snprintf(filter_descr, len, ":fontsize=%d", fontsize);
+  }
+  if (box != -1) {
+    filter_descr += snprintf(filter_descr, len, ":box=%d", box);
+  }
+  if (strcmp(boxcolor, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":boxcolor=%s", boxcolor);
+  }
+  if (strcmp(fontcolor, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":fontcolor=%s", fontcolor);
+  }
+  if (strcmp(fontfile, "") != 0) {
+    filter_descr += snprintf(filter_descr, len, ":fontfile=%s", fontfile);
+  }
+  if (border == 1) {
+    filter_descr +=
+        snprintf(filter_descr, len, ":bordercolor=DarkGray:borderw=1");
+  }
+  if (strcmp(x, "center") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":x=(w-text_w)/2");
+  } else if (strcmp(x, "right") == 0) {
+    // leave 1% margin to the border
+    filter_descr += snprintf(filter_descr, len, ":x=(w-text_w)-w/100");
+  } else if (strcmp(x, "left") == 0) {
+    // leave 1% margin to the border
+    filter_descr += snprintf(filter_descr, len, ":x=w/100");
+  } else { // literal
+    filter_descr += snprintf(filter_descr, len, ":x=%s", x);
+  }
+  if (strcmp(y, "center") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":y=(h-text_h)/2");
+  } else if (strcmp(y, "top") == 0) {
+    // set the same margin for width and height
+    filter_descr += snprintf(filter_descr, len, ":y=w/100");
+  } else if (strcmp(y, "bottom") == 0) {
+    filter_descr += snprintf(filter_descr, len, ":y=(h-text_h)-w/100");
+  } else { // literal
+    filter_descr += snprintf(filter_descr, len, ":y=%s", y);
+  }
+}
```

Comment:
Wouldn't it be better to do on the Elixir side? It's not performance-critical and I guess it could be easier to understand

Reply:
I thought this settings were very ffmpeg-specific (using `h_text` and specific way to pass arguments), so it made sense for me to create it on the native side, elixir side handling only logic. 
But I can move it for clarity, as c notation is indeed messy :)